### DISCUSSION TITLE
[RFC] unify type assertion nodes

### DISCUSSION THREAD

**Discussion by bradzacher** [typescript-eslint/typescript-eslint#6021] [RFCs]
There are two styles for type-assertions:
- `as` assertions (`x as type`)
- "angle bracket" assertions (`<type>x`)

Currently they produce two different, but almost identical ASTs:

```ts
const x = 1 as 1;
```
```jsonc
{
  "type": "TSAsExpression",
  "expression": { /* ... */ },
  "typeAnnotation": { /* ... */ }
}
```

and

```ts
const y = <1>1;
```
```jsonc
{
  "type": "TSTypeAssertion",
  "expression": { /* ... */ },
  "typeAnnotation": { /* ... */ }
}
```

I don't think that there's a reason why these should be separate AST nodes. I _think_ that they're separate nodes because they have different code representations, but I don't know if that's something that matters from the POV  of the AST. Similar to the discussion that's going on around the ESTree representation of optional chaining - there is a difference between the AST and the CST.

From the perspective of ESLint, having two nodes causes pains when writing lint rules, as authors have to remember to add both nodes to selectors or `if` checks. Often they don't, which means rules don't have proper support for TS syntax.

I propose that we unify the two nodes with the following AST:

```ts
interface TSTypeAssertion extends Node {
  type: AST_NODE_TYPES.TSTypeAssertion;
  typeAnnotation: TypeNode;
  expression: Expression;
  kind: 'as' | 'angle-bracket';
}
```

**Comment by JoshuaKGoldberg** [comment]
You know, in retrospect, I don't think there was a need to convert this into a discussion 😄. We like this. And now there's no way to re-open the existing issue. Ah well.

**Comment by bradzacher** [comment]
Worth noting that now we have `TSSatisfiesExpression` as well - so we have 3 nodes for assertions now!

**Comment by JoshuaKGoldberg** [comment]
Moved to #6099 🚀
